### PR TITLE
Ampers: Nora Peters

### PR DESCRIPTION
**Pull request by npeters5**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I read the API docs carefully and tested out a few different requests (q vs. r params) using Postman.
Describe your API Wrapper.  How did you decide on the methods you created? | I created two methods in my RecipeApiWrapper: self.find_recipes to use the 'q' parameter based on user's search terms, to populate index view; and self.get_recipe to make another api request using the 'r' query parameter to look up a specific recipe based on the object's URI, and returning a new Recipe object to display in the show view. 
Describe an edge case or failure case test you wrote for your API Wrapper. | I wrote a test to check that find_recipes returned an empty array if there was no match for the search terms provided.
Explain how VCR aids in testing an API. | Calls to APIs can be expensive, and if the API is down, we're down. VCR helps reduce this external dependency when testing. 
What is the Heroku URL of your deployed application? | https://api-nosher.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/VyU1uMaR/api-muncher-nora

NOTE: I keep getting the error: "uninitialized constant Minitest::Rails::TestUnit (NameError)". I tried all of the things mentioned in Slack and tried troubleshooting a bunch on my own, but I still can't get my tests to run.

### CODE REVIEW

File: test/test_helper.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+ENV["RAILS_ENV"] = "test"
+require File.expand_path("../../config/environment", __FILE__)
+require "rails/test_help"
+
+require "minitest/rails"
+require "minitest/reporters"  # for Colorized output
+
```
Comment (CheezItMan) [inline test/test_helper.rb:7]: You need to have this in here:
```ruby
# MonkeyPatch for This version of Rails :(
if ActionPack::VERSION::STRING >= "5.2.0"
  Minitest::Rails::TestUnit = Rails::TestUnit
end
```

File: app/views/search/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<h3>Results for: "<%= @search_terms %>"</h3>
+<section>
+  <div>
```
Comment (CheezItMan) [inline app/views/search/index.html.erb:3]: Not sure this `div` is really needed here.

File: lib/recipe_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  @recipes_list = []
+
+  def self.find_recipes(search_terms)
+
+    if !@recipes_list.empty?
+      @recipes_list = []
+    end
+
+    encoded_search_terms = URI.encode(search_terms)
```
Comment (CheezItMan) [inline lib/recipe_api_wrapper.rb:18]: There's no checking to ensure that `search_terms` is a string and it's non-empty.

File: app/controllers/search_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+class SearchController < ApplicationController
+
+  def new
+  end
+
+  def index
+    @search_terms = params[:search_terms]
+    @recipes = RecipeApiWrapper.find_recipes(@search_terms).paginate(page: params[:page], per_page: 10)
```
Comment (CheezItMan) [inline app/controllers/search_controller.rb:8]: You need some sort of check to ensure that `@search_terms` is not null. The same applies to `params[:page]`

File: test/controllers/search_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+describe SearchController do
+  it "should get home page" do
+    get root_path
+    must_respond_with :success
+  end
+
+  it "can get recipes index" do
+
+    VCR.use_cassette("recipes") do
+      get recipes_path
```
Comment (CheezItMan) [inline test/controllers/search_controller_test.rb:12]: This test is failing because you're not setting a param here for this path.

```ruby
      get recipes_path, params: {search_terms: query}

```

File: test/controllers/search_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+      query = "chicken"
+      RecipeApiWrapper.find_recipes(query)
```
Comment (CheezItMan) [inline test/controllers/search_controller_test.rb:15]: Why are you making API Wrapper calls directly here?  This should be only making controller calls.

File: test/controllers/search_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    end
+    must_respond_with :success
+  end
+
+  it "should get show" do
+    VCR.use_cassette("recipes") do
+      get recipe_path("7bf4a371c6884d809682a72808da7dc2")
+
+      recipe_id = "7bf4a371c6884d809682a72808da7dc2"
+      RecipeApiWrapper.get_recipe(recipe_id)
```
Comment (CheezItMan) [inline test/controllers/search_controller_test.rb:26]: Again, this is a **controller** test, not a test of the wrapper.

File: test/lib/recipe_api_wrapper_test.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    end
+  end
+
+  describe "get_recipe" do
+
+    before do
+      VCR.use_cassette("get_recipe") do
+        RecipeApiWrapper.find_recipes("chicken")
+        recipe_id = "7bf4a371c6884d809682a72808da7dc2"
+        @recipe = RecipeApiWrapper.find_recipe(recipe_id)
```
Comment (CheezItMan) [inline test/lib/recipe_api_wrapper_test.rb:29]: Should be:

```ruby
@recipe = RecipeApiWrapper.get_recipe(recipe_id)
```

File: test/lib/recipe_api_wrapper_test.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      VCR.use_cassette("recipes") do
+        query = "pork"
+        recipes_list = RecipeApiWrapper.find_recipes(query)
+        recipes_list.must_be_kind_of Array
+      end
+    end
+
+    it "Returns empty array if there are no query matches" do
+      VCR.use_cassette("recipes2") do
+        query = 2429384725434.229384687
```
Comment (CheezItMan) [inline test/lib/recipe_api_wrapper_test.rb:16]: shouldn't this be in quotes?

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | Good number of commits and good commit messages
Comprehension questions | Check, 
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Check, but it would be better if the search results went to the path `/search` 
Semantic HTML | Check 
Errors are reported to the user | No errors reported when there are no search results.  The show page does show a message when the URI is wrong.
API Wrapper to handle the API requests | Check
Controller testing | Lot of issues with controller testing, for one thing you're making calls to the **wrapper** instead of just using the controller in these tests.  You shouldn't be accessing the wrapper directly here.  See my inline notes.
Lib testing | Lots of problems here.  See my inline notes.  
Search Functionality | Check
List Functionality | Check
Show individual item functionality (link to original recipe opens in new tab) | Check
**Styling** | 
Responsive layout | Responsive both on the index and show pages. 
List View shows 10 items at a time/pagination | Check
The app is styled to create an attractive user interface | 
**API Features** | 
The App attributes Edaman | **MISSING**
The VCR cassettes do not contain the API key | Check
**External Resources** | 
Link to Trello Board | Check
Link to deployed app on Heroku | Check
**Overall** | You did well with the API, but you have some serious problems with your testing.  Please review my notes and let me know where I can clarify.  You won't have any more opportunities for Rails testing and like you to take some time and see if you can see where you went wrong and how to do the testing properly.  I'm happy to go through it with you if needed.  You hit the API goals for the project, and had trouble with the testing, but you did well with the most visible portions of the project.  I hope you enjoyed Rails and learn to love JavaScript starting next week!

**Review by CheezItMan — COMMENTED** [review]
The most obvious thing is you missed the "monkey" patch which we had to include from the SlackAPi project.  That might explain your testing problems.
